### ISSUE TITLE
Re-authenticate when refresh token expires

### ISSUE THREAD

**Issue by atwoosnam** [vmware-tanzu/tanzu-toolkit-for-visual-studio#37]
The CF CLI manages 2 auth tokens (it stores them in `~/.cf/config.json`):

1. An **Access Token** with a 2-hour lifespan. When this token expires mid-session, the CF CLI requests a new one from the CF API using:...
2. A **Refresh Token** with a 2-week lifespan. When this token expires, the CF CLI requires re-authentication (from [what I understand](https://vmware.slack.com/archives/C0555EXQJ/p1574962374093300?thread_ts=1574918070.093100&cid=C0555EXQJ))

Our extension currently only performs an authentication request (i.e. retrieval of a new refresh token) when connecting to a new "Cloud Foundry Instance" for the first time. This means that we can expect users of our extension to encounter issues after 2 weeks of using the same "Cloud connection" if they don't manually re-login (and nothing about our extension makes that clear at the moment).

**Comment by atwoosnam** [comment]
Possible solution:

1. After authenticating for the first time, read the CF CLI config file & record the value of the refresh token
    - (unfortunately, there might not be a better way to read the refresh token; there's no CF CLI method that will return the value, and the refresh token is obfuscated when printing all http info using `cf login -v`)


2. In the GetOAuthToken method, before checking the access token's expiration time, check the refresh token's expiration time

3. If the refresh token hasn't yet expired, proceed to check access token validity etc.
4. If the refresh token has expired, prompt re-authentication (this might require a bit of refactor to prevent CfCliService from invoking CloudFoundryService methods)

**Comment by atwoosnam** [comment]
We could _guess_ where the CF CLI config file lives (`~/.cf` is the default, so that'd be a decent guess), but it'll be safer to read the refresh token from config.json file once this VS extension has a dedicated location for that file (see #89).

**Comment by atwoosnam** [comment]
CF CLI METHODS
- valid AT, valid RT => request succeeds
- invalid AT, valid RT => CF CLI uses RT to get new AT & retries request
- valid AT, invalid RT => request succeeds
- invalid AT, invalid RT => CF CLI tries to use RT to get a new AT & fails with exit code 1. 
  - If StdErr == "The token expired, was revoked, or the token ID is incorrect. Please log back in to re-authenticate.", prompt login
  - Else, proceed with default error (issue is not an invalid RT)

CF API CLIENT METHODS:
- valid AT, valid RT => request succeeds
- invalid AT, valid RT => method retried in CloudFoundryService with fresh AT
- valid AT, invalid RT => request succeeds
- invalid AT, invalid RT => method retried once in CloudFoundryService, fails again. After failed retry, read RT from CF CLI config file & check for validity. 
  - If RT is expired, prompt login
  - Else, throw default error (this *assumes* that the RT is still valid, which could be an issue if the RT became invalid *before* the end of its precribed lifetime)


Sample login window title:
Your session has expired, please log back in to re-authenticate.

**Comment by atwoosnam** [comment]
Do we have to worry about canceling tasks when the refresh token expires in a scenario like this?

- refresh poller initiates 10 refresh tasks: "taskA" - "taskJ"
- taskA completes successfully
- taskB completes successfully
- refresh token expires (but access token is still valid)
- taskC completes successfully
- taskD completes successfully
- taskE completes successfully
- access token expires
- taskF fails the first time due to an expired access token & tries to get a fresh access token using the (expired) refresh token
- some (not yet implemented) process recognizes an expired refresh token & prompts the user to log back in for re-authentication
- TaskG fails the first time due to an expired access token, ... then what? Re-prompts the user to log back in for re-authentication due to an expired refresh token?
- TaskH fails the first time due to an expired access token, ... then what? Re-prompts the user to log back in for re-authentication due to an expired refresh token?
- TaskI fails the first time due to an expired access token, ... then what? Re-prompts the user to log back in for re-authentication due to an expired refresh token?
- TaskJ fails the first time due to an expired access token, ... then what? Re-prompts the user to log back in for re-authentication due to an expired refresh token?
- ... those are a lot of unnecessary re-prompts ...

**Comment by atwoosnam** [comment]
New idea: 
Let's keep track of whether or not a `CloudFoundryInstance` shown in the CloudExplorer window is authenticated & have the refresh poller check this value before it starts each round of background refreshes.
When first logging in to a CF instance, mark it as authenticated.
Whenever a CF operation needs to get a fresh access token but the refresh token is expired, mark that CF instance as NOT authenticated.
Whenever the refresh poller discovers that a CF instance is not authenticated (specifically, when it tries to run `CloudExplorerViewModel.RefreshAllItems()`), display some sort of notification INSIDE the Cloud Explorer tool window that indicates re-authentication is required.

**Comment by atwoosnam** [comment]
`CfCliService.GetOAuthToken` is now throwing an `InvalidRefreshTokenException` whenever it gets this response as StdErr from the `CmdProcessService`: "_The token expired, was revoked, or the token ID is incorrect. Please log back in to re-authenticate._"

All 6 CloudFoundryService methods which use `CfCliService.GetOAuthToken` (get orgs/spaces/apps + stop/start/delete app) are now catching any `InvalidRefreshTokenException`s & responding by returning a `DetailedResult` with `FailureType` == `InvalidRefreshToken`.

I've tested this in the debugger & seen it work for GetOrgsAsync! I'm assuming the same is true for the other 5 methods, since the code is identical.

TODO:
- have the cloud explorer view model monitor the `Authenticated` status of each CloudFoundryInstance & take appropriate action when it notices that one has become unauthenticated.
- fix refresh flow so that no refresh attempt is made when the CF in question is marked as not authenticated.
- make sure no modals (error windows) pop up because of an invalid refresh token during the background refresh task.
- investigate tree view behavior when tokens are both invalid during a refresh -- should the methods for getting orgs/spaces/apps return empty lists (e.g. report "no orgs")? Or should they return whatever was in place before the refresh to simulate a no-op (i.e. leave the tree view visually unchanged)?

**Comment by atwoosnam** [comment]
TODO:
- Change all predicates in CloudExplorerViewModel to return false if not Authenticated

**Comment by atwoosnam** [comment]
De-authentication via automatic refresh behavior has been verified for these methods:

- [ ] `CfInstanceViewModel.RefreshChildren`
- [ ] `OrgViewModel.RefreshChildren`
- [x] `SpaceViewModel.RefreshChildren`

**Comment by atwoosnam** [comment]
Just pushed a series of commits to `refresh-token-expiration-#37` (kept previous unpolished progress in a separate branch on the remote: `refresh-token-expiration-#37-old`). These changes make it so that any "active" attempt to get orgs/spaces/apps which fails because of an invalid refresh token will display an error pop-up explaining the issue & set `CloudExplorerViewModel.AuthenticationRequired` to true. Any "passive" attempts to fetch children for cfs/orgs/spaces will not display any pop-ups but will set the `AuthenticationRequired` flag the same way. Whenever `CloudExplorerViewModel.AuthenticationRequired` is set to true, the window displays a button (next to the refresh button for now) which, when clicked, will remove the topmost CF (there should only ever be 1) & display the login window again. A successful login will reset `AuthenticationRequired` to false, clearing the auth warning button.

TODO:
- remove unused prop on CloudFoundryInstance: `Authenticated` (no longer need this value as it's kept track of on the CloudExplorerViewModel now that I've replaced the auth-polling system with an auth feedback mechanism from each TreeViewItemViewModel back to the CloudExplorerViewModel).
- add support for other methods that may require re-authentication, like app deployment or getting app logs 
  - _note: some of these may not use `CfCliService.GetOAuthToken` (e.g. if they happen via the CF CLI entirely and not the CfApiClient), so we'll need another mechanism for detecting when the refresh token is invalid (maybe we can scan the StdErr for the same failure message?)_
- optionally: condense all logic having to do with verifying auth stuff into a single place?

**Comment by atwoosnam** [comment]
Here are the CF CLI methods we're currently using:

1.  GetOAuthToken (6 usages in CloudFoundryService)
2. TargetApi (1 usage in CloudFoundryService.ConnectToCFAsync)
3. AuthenticateAsync (1 usage in CloudFoundryService.ConnectToCFAsync)
4. PushAppAsync (1 usage in CloudFoundryService.DeployAppAsync)
5. GetApiVersion (1 usage in CloudFoundryService.MatchCliVersionToApiVersion)
6. GetRecentAppLogs (1 usage in CloudFoundryService.GetRecentLogs)
7. ClearCachedAccessToken (6 usages in CloudFoundryService)

ExecuteCfCliCommand is only in use within CfCliService itself; this method is probably safe to make private.

All other CF CLI methods are tested but not in use (GetOrgsAsync, GetSpacesAsync, TargetOrg, TargetSpace, GetAppsAsync, StopAppByNameAsync, StartAppByNameAsync, DeleteAppByNameAsync)

Fore each of those 7 used methods, here's a table showing their underlying CF CLI methods + whether or not those methods fail due to an invalid refresh token:

|CF CLI Method |Token-dependent |Token-independent |
|--- | :---: | :---:|
|1. `oauth-token`|✅||
|2. `api`||✅|
|3. `auth`||
|4. `push`||
|5. `api`||
|6. `logs`||
|7. **N/A** (doesn't use CF CLI)||

**Comment by atwoosnam** [comment]
The comment above shows that `oauth-token`, `push`, and `logs` are the only 3 CF CLI methods in use by our extension that could encounter an error due to an invalid refresh token. 

- `oauth-token` is now covered (as of the changes pushed up to the `refresh-token-expiration-#37` branch last night).
- `push` needs protection. Since this is an "active" operation within our extension, an error notification should pop up if a deployment can't be completed due to an invalid refresh token.
- `logs` needs protection. Since this is an "active" operation within our extension, an error notification should pop up if a deployment can't be completed due to an invalid refresh token.

**Comment by atwoosnam** [comment]
Turns out we can't completely ignore the other methods in CfCliService (GetOrgsAsync, GetSpacesAsync, TargetOrg, TargetSpace, GetAppsAsync, StopAppByNameAsync, StartAppByNameAsync, and DeleteAppByNameAsync) because some are still in use within "used" methods on CfCliService.

E.g. `PushAppAsync` uses `TargetOrg` & `TargetOrg` is at risk of encountering an error due to an invalid refresh token.

I've added extra rows to the table from above to capture the importance of guarding against invalid refresh token errors for those remaining methods:
